services/sharedchannel/sync_send.go

```diff
@@ -329,10 +330,22 @@ func (scs *Service) updateForRemote(task syncTask, rc *model.RemoteCluster) erro
 	}
 	msg := model.NewRemoteClusterMsg(TopicSync, b)
 
+	if scs.server.GetLogger().IsLevelEnabled(mlog.LvlSharedChannelServiceMessagesOutbound) {
+		scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceMessagesOutbound, "outbound message",
+			mlog.String("remote", rc.DisplayName),
+			mlog.Int64("NextSyncAt", scr.NextSyncAt),
+			mlog.String("msg", string(b)),
+		)
+	}
+
 	ctx, cancel := context.WithTimeout(context.Background(), remotecluster.SendTimeout)
 	defer cancel()
 
+	var wg sync.WaitGroup
```

Comment:
So in order for this `wg` to be released the callback function has to be called. So if in `SendMsg` we get a panic, we will `recover()` but we will never call the callback function, essentially deadlocking the caller, right?

Now, the `sendFrameToRemote` looks pretty safe but just highlighting this in case that we are missing anything.

Reply:
Correct, we need the callback called, which is pretty safe IMO.  However I refactored slightly to force the callback even on a panic.  https://github.com/mattermost/mattermost-server/pull/17125/commits/0e96676678c5dd369f44f448a6415c3943dc00f0